From c4dc4a921016047d4dbc4a0f2639b6aacbc34572 Mon Sep 17 00:00:00 2001
From: Eric Yang <eric.yang.wiwynn@gmail.com>
Date: Thu, 15 Aug 2024 17:27:50 +0800
Subject: [PATCH] meta-facebook: yosemite4: Prevent BMC reboot by
 disabling/enabling i3c hub

During hot service, disable and enable the i3c hub to prevent potential BMC reboots caused by noise from slot plug-in.

This change addresses the issue where enabling the i3c port during slot plug-in can generate electrical noise due to leakage.

These noises, when sent to the BMC by the i3c hub, cause interrupt storming, leading to the BMC CPU getting stuck and unexpected reboots.

Tested:
- Verified that the i3c hub is correctly disabled and enabled during hot service -pass.
- Performed 60 plug-in and unplug operations, ensuring only one slot was operated at a time. No BMC reboots were observed -pass.

Change-Id: I030daf8b4b68a43204847eeb2153757ccb18292f
Signed-off-by: Eric Yang <eric.yang.wiwynn@gmail.com>
---
 .../phosphor-gpio-monitor/disable-i3c-hub     | 50 +++++++++++++
 .../disable-i3c-hub@.service                  |  7 ++
 .../gpio/phosphor-gpio-monitor/enable-i3c-hub | 50 +++++++++++++
 .../enable-i3c-hub@.service                   |  7 ++
 ...yosemite4-phosphor-multi-gpio-monitor.json | 72 ++++++++++++++-----
 .../gpio/phosphor-gpio-monitor_%.bbappend     | 10 +++
 6 files changed, 180 insertions(+), 16 deletions(-)
 create mode 100644 meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub
 create mode 100644 meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub@.service
 create mode 100644 meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub
 create mode 100644 meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub@.service

diff --git a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub
new file mode 100644
index 00000000000..b8268e86685
--- /dev/null
+++ b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+disable_i3c_hub()
+{
+    local unplugged_slot="$1"
+    local hub_path
+    local offset_file
+    local access_file
+    local current_port
+    local mask=1
+
+    if [ "$unplugged_slot" -lt 5 ]; then
+        hub_path="/sys/kernel/debug/i3c-hub-0-*/reg"
+        echo "Slot $unplugged_slot on i3c hub 0."
+    else
+        hub_path="/sys/kernel/debug/i3c-hub-1-*/reg"
+        echo "Slot $unplugged_slot on i3c hub 1."
+    fi
+
+    for file in $hub_path/{offset,access}; do
+        if [[ $file == *"/offset" ]]; then
+            offset_file="$file"
+        elif [[ $file == *"/access" ]]; then
+            access_file="$file"
+        fi
+    done
+
+    echo "Unlock i3c hub register."
+    echo 16 > "$offset_file"
+    echo 105 > "$access_file"
+
+    echo "Disable slot${unplugged_slot} i3c port."
+    echo 18 > "$offset_file"
+    current_port=$(cat "$access_file")
+    if [ "$unplugged_slot" -gt 4 ]; then
+        ((unplugged_slot=unplugged_slot-4))
+    fi
+    mask=$((mask << (--unplugged_slot)))
+    mask=$((~mask))
+    current_port=$((current_port & mask))
+    echo $current_port > "$access_file"
+
+    echo "Lock i3c hub register."
+    echo 16 > "$offset_file"
+    echo 0 > "$access_file"
+
+    exit 0
+}
+
+disable_i3c_hub "$1"
diff --git a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub@.service b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub@.service
new file mode 100644
index 00000000000..2ffa66b14d2
--- /dev/null
+++ b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/disable-i3c-hub@.service
@@ -0,0 +1,7 @@
+[Unit]
+Description=slot-disable-i3c-hub:%i
+
+[Service]
+Type=oneshot
+ExecStart=/usr/libexec/phosphor-gpio-monitor/disable-i3c-hub %i
+SyslogIdentifier=slot-disable-i3c-hub %i
diff --git a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub
new file mode 100644
index 00000000000..f57c7654e95
--- /dev/null
+++ b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+enable_i3c_hub()
+{
+    local plugged_slot="$1"
+    local hub_path
+    local offset_file
+    local access_file
+    local current_port
+    local mask=1
+
+    sleep 2
+    if [ "$plugged_slot" -lt 5 ]; then
+        hub_path="/sys/kernel/debug/i3c-hub-0-*/reg"
+        echo "Slot${plugged_slot} on i3c hub 0."
+    else
+        hub_path="/sys/kernel/debug/i3c-hub-1-*/reg"
+        echo "Slot${plugged_slot} on i3c hub 1."
+    fi
+
+    for file in $hub_path/{offset,access}; do
+        if [[ $file == *"/offset" ]]; then
+            offset_file="$file"
+        elif [[ $file == *"/access" ]]; then
+            access_file="$file"
+        fi
+    done
+
+    echo "Unlock i3c hub register."
+    echo 16 > "$offset_file"
+    echo 105 > "$access_file"
+
+    echo "Enable slot${plugged_slot} i3c port."
+    echo 18 > "$offset_file"
+    current_port=$(cat "$access_file")
+    if [ "$plugged_slot" -gt 4 ]; then
+        ((plugged_slot=plugged_slot-4))
+    fi
+    mask=$((mask << (--plugged_slot)))
+    current_port=$((current_port | mask))
+    echo $current_port > "$access_file"
+
+    echo "Lock i3c hub register."
+    echo 16 > "$offset_file"
+    echo 0 > "$access_file"
+
+    exit 0
+}
+
+enable_i3c_hub "$1"
diff --git a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub@.service b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub@.service
new file mode 100644
index 00000000000..4d9c0839a1d
--- /dev/null
+++ b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/enable-i3c-hub@.service
@@ -0,0 +1,7 @@
+[Unit]
+Description=slot-enable-i3c-hub:%i
+
+[Service]
+Type=oneshot
+ExecStart=/usr/libexec/phosphor-gpio-monitor/enable-i3c-hub %i
+SyslogIdentifier=slot-enable-i3c-hub %i
diff --git a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/yosemite4-phosphor-multi-gpio-monitor.json b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/yosemite4-phosphor-multi-gpio-monitor.json
index ab898a90287..f1d5941e67d 100644
--- a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/yosemite4-phosphor-multi-gpio-monitor.json
+++ b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor/yosemite4-phosphor-multi-gpio-monitor.json
@@ -163,10 +163,15 @@
         "Name": "PRSNT_SB_SLOT1_BMC_N",
         "ChipId": "0",
         "GpioNum": 44,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@1.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@1.service"]
+                "slot-hot-plug@1.service",
+                "enable-i3c-hub@1.service"
+            ]
         },
         "Continue": true
     },
@@ -174,10 +179,15 @@
         "Name": "PRSNT_SB_SLOT2_BMC_N",
         "ChipId": "0",
         "GpioNum": 45,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@2.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@2.service"]
+                "slot-hot-plug@2.service",
+                "enable-i3c-hub@2.service"
+            ]
         },
         "Continue": true
     },
@@ -185,10 +195,15 @@
         "Name": "PRSNT_SB_SLOT3_BMC_N",
         "ChipId": "0",
         "GpioNum": 46,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@3.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@3.service"]
+                "slot-hot-plug@3.service",
+                "enable-i3c-hub@3.service"
+            ]
         },
         "Continue": true
     },
@@ -196,10 +211,15 @@
         "Name": "PRSNT_SB_SLOT4_BMC_N",
         "ChipId": "0",
         "GpioNum": 47,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@4.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@4.service"]
+                "slot-hot-plug@4.service",
+                "enable-i3c-hub@4.service"
+            ]
         },
         "Continue": true
     },
@@ -207,10 +227,15 @@
         "Name": "PRSNT_SB_SLOT5_BMC_N",
         "ChipId": "0",
         "GpioNum": 132,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@5.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@5.service"]
+                "slot-hot-plug@5.service",
+                "enable-i3c-hub@5.service"
+            ]
         },
         "Continue": true
     },
@@ -218,10 +243,15 @@
         "Name": "PRSNT_SB_SLOT6_BMC_N",
         "ChipId": "0",
         "GpioNum": 133,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@6.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@6.service"]
+                "slot-hot-plug@6.service",
+                "enable-i3c-hub@6.service"
+            ]
         },
         "Continue": true
     },
@@ -229,10 +259,15 @@
         "Name": "PRSNT_SB_SLOT7_BMC_N",
         "ChipId": "0",
         "GpioNum": 134,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@7.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@7.service"]
+                "slot-hot-plug@7.service",
+                "enable-i3c-hub@7.service"
+            ]
         },
         "Continue": true
     },
@@ -240,10 +275,15 @@
         "Name": "PRSNT_SB_SLOT8_BMC_N",
         "ChipId": "0",
         "GpioNum": 135,
-        "EventMon": "FALLING",
+        "EventMon": "BOTH",
         "Targets": {
+            "RISING": [
+                "disable-i3c-hub@8.service"
+            ],
             "FALLING": [
-                "slot-hot-plug@8.service"]
+                "slot-hot-plug@8.service",
+                "enable-i3c-hub@8.service"
+            ]
         },
         "Continue": true
     }
diff --git a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor_%.bbappend b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor_%.bbappend
index 112b3424ac4..5507f56680c 100644
--- a/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor_%.bbappend
+++ b/meta-facebook/meta-yosemite4/recipes-phosphor/gpio/phosphor-gpio-monitor_%.bbappend
@@ -16,6 +16,10 @@ SRC_URI += "file://yosemite4-phosphor-multi-gpio-monitor.json \
             file://slot-hot-plug@.service \
             file://rescan-wf-bic \
             file://rescan-wf-bic@.service \
+            file://enable-i3c-hub \
+            file://enable-i3c-hub@.service \
+            file://disable-i3c-hub \
+            file://disable-i3c-hub@.service \
             "
 
 RDEPENDS:${PN}:append = " bash"
@@ -31,6 +35,8 @@ SYSTEMD_SERVICE:${PN} += " \
     remove-nic-endpoint-slot@.service \
     rescan-wf-bic@.service \
     reconfig-net-interface@.service \
+    enable-i3c-hub@.service \
+    disable-i3c-hub@.service \
     "
 
 SYSTEMD_AUTO_ENABLE = "enable"
@@ -47,11 +53,15 @@ do_install:append:() {
     install -m 0644 ${WORKDIR}/reconfig-net-interface@.service  ${D}${systemd_system_unitdir}/reconfig-net-interface@.service
     install -m 0644 ${WORKDIR}/remove-nic-endpoint-slot@.service ${D}${systemd_system_unitdir}/
     install -m 0644 ${WORKDIR}/rescan-wf-bic@.service ${D}${systemd_system_unitdir}/
+    install -m 0644 ${WORKDIR}/enable-i3c-hub@.service ${D}${systemd_system_unitdir}/
+    install -m 0644 ${WORKDIR}/disable-i3c-hub@.service ${D}${systemd_system_unitdir}/
     install -d ${D}${libexecdir}/${PN}
     install -m 0755 ${WORKDIR}/probe-slot-device ${D}${libexecdir}/${PN}/
     install -m 0755 ${WORKDIR}/reconfig-net-interface ${D}${libexecdir}/${PN}/
     install -m 0755 ${WORKDIR}/rescan-fru-device ${D}${libexecdir}/${PN}/
     install -m 0755 ${WORKDIR}/rescan-wf-bic ${D}${libexecdir}/${PN}/
+    install -m 0755 ${WORKDIR}/enable-i3c-hub ${D}${libexecdir}/${PN}/
+    install -m 0755 ${WORKDIR}/disable-i3c-hub ${D}${libexecdir}/${PN}/
     install -d ${D}/${bindir}
     install -m 0755 ${WORKDIR}/configure-nic-mctp-endpoint.sh ${D}/${bindir}/
 }
-- 
2.25.1

